File: stepic.com/week1/fib.hs
module Fib where

{-fibonacci n | n == (-1) = 1
            | n == 0 = 0
            | n == 1 = 1
            | n > 1 = fibonacci (n - 1) + fibonacci (n - 2)
            | otherwise = fibonacci (n + 2) - fibonacci (n + 1)


fibonacciAccum  n = (helper [] (n - 1) fibonacciAccum) + (helper [] (n - 2) fibonacciAccum)

helper accumList n func | n == 0 = 0
                        | n == 1 = 1
                        | (length accumList) >= n = accumList !! ((length accumList) - n)
                        | otherwise = ((func n ): accumList) !! 0 -- Выполняем функцию > добавляем в лист > выплевываем это число
-}

fib2 n = helper1 n 0 1

helper1 n a b | n == 0 = a
              | n > 0 = helper1 (n-1) b (a + b) 
              | n < 0 = helper1 (n+1) b (a - b)


fib3 n = 
    let helper n a b | n == 0 = a
                     | n > 0 = helper (n-1) b (a + b)
    in helper n 0 1

File: stepic.com/week1/fib.py
def helper(some_list, index, callback):
	# import pdb; pdb.set_trace()
	if index <= 1:
		return index
	try:
		return some_list[len(some_list) - index]
	except:
		cur_num = callback(index, some_list)
		some_list.insert(0, cur_num)
		return cur_num


def fib2(n, alist):
	# if n <= 1:
		# return n
	# else:
	return helper(alist, n-1, fib2) + helper(alist, n-2, fib2)


def fib3(n):
	return helper1(n-1, 0, 1)


def helper1(index, last, cur):
	if index == 0 :
		return last

	helper1(index-1, last)

print(fib2(30, []))